### ISSUE TITLE
Suggestion to remove "react-optimize" babel option

### ISSUE THREAD

**Issue by doortts** [chentsulin/electron-react-boilerplate#1634]
I found that it may occur some serious and hard to find problems 

![image](https://user-images.githubusercontent.com/520523/42724722-99366a88-87b2-11e8-8d21-269df6b4d584.png)

https://github.com/chentsulin/electron-react-boilerplate/blob/master/.babelrc#L13

in some circumstance, babel will make errors like

>  We don't know what to do with this node type. We were previously a Statement but we can't fit in here?

which is caused by https://github.com/jamiebuilds/babel-react-optimize/tree/master/packages/babel-plugin-transform-react-pure-class-to-function

I had to waste lots of time to find it.

**Comment by amilajack** [comment]
What is the error that you got?

**Comment by doortts** [comment]
I couldn't find a particular form of case, but from time to time it occurrs in very normal react containers.

main error message is 

> We don't know what to do with this node type. We were previously a Statement but we can't fit in here?

Following is the error I faced yesterday.

```
ERROR in ./app/containers/ListContainer/index.js
[1] Module build failed (from ./node_modules/babel-loader/lib/index.js):
[1] Error: /Users/doortts/repos/new-dc/app/containers/ListContainer/index.js: We don't know what to do with this node type. We were previously a Statement but we can't fit in here?
[1]     at NodePath.insertAfter (/Users/doortts/repos/new-dc/node_modules/babel-traverse/lib/path/modification.js:175:13)
[1]     at NodePath.replaceWithMultiple (/Users/doortts/repos/new-dc/node_modules/babel-traverse/lib/path/replacement.js:82:8)
[1]     at PluginPass.Class (/Users/doortts/repos/new-dc/node_modules/babel-plugin-transform-react-pure-class-to-function/lib/index.js:96:14)
[1]     at newFn (/Users/doortts/repos/new-dc/node_modules/babel-traverse/lib/visitors.js:276:21)
[1]     at NodePath._call (/Users/doortts/repos/new-dc/node_modules/babel-traverse/lib/path/context.js:76:18)
[1]     at NodePath.call (/Users/doortts/repos/new-dc/node_modules/babel-traverse/lib/path/context.js:48:17)
[1]     at NodePath.visit (/Users/doortts/repos/new-dc/node_modules/babel-traverse/lib/path/context.js:105:12)
[1]     at TraversalContext.visitQueue (/Users/doortts/repos/new-dc/node_modules/babel-traverse/lib/context.js:150:16)
[1]     at TraversalContext.visitSingle (/Users/doortts/repos/new-dc/node_modules/babel-traverse/lib/context.js:108:19)
[1]     at TraversalContext.visit (/Users/doortts/repos/new-dc/node_modules/babel-traverse/lib/context.js:192:19)
[1]     at Function.traverse.node (/Users/doortts/repos/new-dc/node_modules/babel-traverse/lib/index.js:114:17)
[1]     at NodePath.visit (/Users/doortts/repos/new-dc/node_modules/babel-traverse/lib/path/context.js:115:19)
[1]     at TraversalContext.visitQueue (/Users/doortts/repos/new-dc/node_modules/babel-traverse/lib/context.js:150:16)
[1]     at TraversalContext.visitMultiple (/Users/doortts/repos/new-dc/node_modules/babel-traverse/lib/context.js:103:17)
[1]     at TraversalContext.visit (/Users/doortts/repos/new-dc/node_modules/babel-traverse/lib/context.js:190:19)
[1]     at Function.traverse.node (/Users/doortts/repos/new-dc/node_modules/babel-traverse/lib/index.js:114:17)
[1]     at NodePath.visit (/Users/doortts/repos/new-dc/node_modules/babel-traverse/lib/path/context.js:115:19)
[1]     at TraversalContext.visitQueue (/Users/doortts/repos/new-dc/node_modules/babel-traverse/lib/context.js:150:16)
[1]     at TraversalContext.visitSingle (/Users/doortts/repos/new-dc/node_modules/babel-traverse/lib/context.js:108:19)
[1]     at TraversalContext.visit (/Users/doortts/repos/new-dc/node_modules/babel-traverse/lib/context.js:192:19)
[1]     at Function.traverse.node (/Users/doortts/repos/new-dc/node_modules/babel-traverse/lib/index.js:114:17)
[1]     at traverse (/Users/doortts/repos/new-dc/node_modules/babel-traverse/lib/index.js:79:12)
[1]     at File.transform (/Users/doortts/repos/new-dc/node_modules/babel-core/lib/transformation/file/index.js:516:35)
[1]     at /Users/doortts/repos/new-dc/node_modules/babel-core/lib/transformation/pipeline.js:50:19
[1]     at File.wrap (/Users/doortts/repos/new-dc/node_modules/babel-core/lib/transformation/file/index.js:532:16)
[1]     at Pipeline.transform (/Users/doortts/repos/new-dc/node_modules/babel-core/lib/transformation/pipeline.js:47:17)
[1]     at transpile (/Users/doortts/repos/new-dc/node_modules/babel-loader/lib/index.js:50:20)
[1]     at /Users/doortts/repos/new-dc/node_modules/babel-loader/lib/fs-cache.js:118:18
    at ReadFileContext.callback (/Users/doortts/repos/new-dc/node_modules/babel-loader/lib/fs-cache.js:31:21)
[1]     at FSReqWrap.readFileAfterOpen [as oncomplete] (fs.js:354:13)
[1]  @ ./app/containers/Home/index.js 30:21-56
[1]  @ ./app/routes.js
[1]  @ ./app/containers/Root.js
[1]  @ ./app/index.js
[1] Child extract-text-webpack-plugin node_modules/extract-text-webpack-plugin/dist node_modules/css-loader/index.js??ref--6-1!node_modules/sanitize.css/sanitize.css:
[1]        2 modules
[1] npm ERR! code ELIFECYCLE
```

ListContainer is nothing special.
![image](https://user-images.githubusercontent.com/520523/42733359-29c37b46-886b-11e8-8ce3-f229aee3442d.png)

**Comment by doortts** [comment]
```
at PluginPass.Class (/Users/doortts/repos/new-dc/node_modules/babel-plugin-transform-react-pure-class-to-function/lib/index.js:96:14)
```
That was a only clue which one was the problem.

babel-plugin-transform-react-pure-class-to-function

is derived from react-optimize plugin.

**Comment by Slapbox** [comment]
I don't think `react-optimize` was ever ported over to `Babel 7`. If I'm right, it might not be unreasonable to remove it, since I assume a transition to that version is in the near future.
----
**Off topic, but tangentially related** @amilajack I have all the code needed to upgrade to `Babel 7`, but it's mixed into proprietary code which is what's delayed me from submitting a completed transition to `Babel 7` and `Webpack 4`. Webpack was the toughest part. Since I see that's been upgraded, I can look into modifying my configurations for Recollectr to work with the base boilerplate again.

Are you comfortable with the stability of version 7 yet? If so I can look into that. 

One thing that may be broken in any PR I submit is the internal scripts folder. Not sure when/how I lost that since I don't use it anymore, but in any case, I broke it.

**Comment by doortts** [comment]
In addition, https://github.com/jamiebuilds/babel-react-optimize doesn't accept issues. So, I couldn't submit this problem to the react-optimize repository.

**Comment by amilajack** [comment]
> Are you comfortable with the stability of version 7 yet? If so I can look into that.

The unsaid standard of erb is that we don't support experimental versions of build tools like webpack and babel until a few weeks or months after they've been released. The idea behind this is that they'll be at least a little battle tested before we support them. Linting tools are okay with being experimental because they cause infra issues instead of app related issues, which are much worse.

**Comment by amilajack** [comment]
@doortts can you provide a fork that demonstrates this issue? I'd love to investigate this more.

**Comment by vikr01** [comment]
This will be resolved in #1763.
